src/read.c

```diff
@@ -2723,8 +2731,8 @@ UInt ReadEvalFile(Obj * evalResult)
 */
 void ReadEvalError(void)
 {
-    STATE(PtrBody)  = PTR_BAG(BODY_FUNC(CURR_FUNC()));
-    STATE(PtrLVars) = PTR_BAG(STATE(CurrLVars));
+    GAP_ASSERT(STATE(PtrBody) == PTR_BAG(BODY_FUNC(CURR_FUNC())));
+    GAP_ASSERT(STATE(PtrLVars) == PTR_BAG(STATE(CurrLVars)));
```

Comment:
Actually, I just saw that `CURR_FUNC` actually contains this check:
```C
EXPORT_INLINE Obj CURR_FUNC(void)
{
    GAP_ASSERT(IS_LVARS_OR_HVARS(STATE(CurrLVars)));
    GAP_ASSERT(STATE(PtrLVars) == PTR_BAG(STATE(CurrLVars)));
    return FUNC_LVARS_PTR(STATE(PtrLVars));
}
```
So clearly at least the second assertion must have always been satisfied.

Moreover, we do this in `vars.c`:
```
static void VarsAfterCollectBags(void)
{
  if (STATE(CurrLVars))
    {
      STATE(PtrLVars) = PTR_BAG( STATE(CurrLVars) );
      STATE(PtrBody)  = PTR_BAG( BODY_FUNC( CURR_FUNC() ) );
    }
}
```
This is called by a GC, so that can't mess up `PtrBody`. So really the only way `PtrBody` could be "wrong" is if we modify it then call `ReadEvalError` *before* any GC has a chance of happening. So where is `PtrBody` changed? First, by `SWITCH_TO_*_LVARS` but those three functions preserve the invariants. That leaves a single other place modifying it, in `NewStatOrExpr`:
```C
...
    Obj body = BODY_FUNC(CURR_FUNC());
    UInt bodySize = SIZE_BAG(body);
    if (bodySize == 0)
        bodySize = CS(OffsBody);
    while (bodySize < CS(OffsBody))
        bodySize *= 2;
    ResizeBag(body, bodySize);
    STATE(PtrBody) = PTR_BAG(body);
...
```
As you can see the invariant is preserved, and in fact the assignment unnecessary. I'll change that into an assertion, too.

Reply:
Actually the assignment is *not* unnecessary, simply because ResizeBag can of course move a bag even without a GC taking place. But I now added an assertion *before* the `ResizeBag` to verify the invariant.